### PR TITLE
Improve unsafety (Fixes #36692 and #36024)

### LINKED ISSUE #36024: include_vars is not correctly expanding included expressions
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
`include_vars`

##### ANSIBLE VERSION
```
ansible 2.4.2.0
  config file = /Users/alex/works/sysadm/ansible-study/ansible.cfg
  configured module search path = [u'/Users/alex/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/alex/VEnvs/virtualenv/ansible2.4/lib/python2.7/site-packages/ansible
  executable location = /Users/alex/VEnvs/virtualenv/ansible2.4/bin/ansible
  python version = 2.7.10 (default, Feb  7 2017, 00:08:15) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.34)]
```

##### OS / ENVIRONMENT
macOS, linux

##### SUMMARY
When some of vars in include file are not defined and are used in expressions, results from inclusion are incorrect and inconsistent: sometimes expressions are not expanded and are included verbatim, sometimes exceptions are raised even with safety defaults etc. Additionally, results are different depending on `register: results` attribute and `name: varname` parameter of the task.
Most probably related to [bug 21088](https://github.com/ansible/ansible/issues/21088).

##### STEPS TO REPRODUCE
Example inventory (to use `hostvars` below we need one):

```
# dummy group for -l localhost to work
[localhosts]
localhost     ansible_ssh_host=localhost
```

File with vars to include: expression for `var5` has an error, in `var4` we will try to provide 
default value for bad variable

```yaml
var1: "a"
# change to "var2" to fix
no_var2: "b"

var3: "{{ 'just ' + var1 }}"
var4: "{{ var1 + ' and ' + var2|default('<no second>') }}"
var5: "{{ var1 + ' and ' + var2 }}"
```

And, finally, an example playbook

```yaml
# ansible-playbook -l localhost include_expansion_bug.yml
- hosts: localhost
  gather_facts: false

  tasks:

  - name: include vars from file (mostly ok, broken 'default')
    include_vars:
      file: "include_expansion_vars.yml"

  - name: print var4 (direct), expecting "a and <c>"
    debug: var=var4

  - name: print var4 (host vars), expecting "a and <c>"
    debug: var=hostvars[inventory_hostname].var4

  - name: print var4 (msg), expecting "a and <c> == a and <c>"
    debug: msg="{{ hostvars[inventory_hostname].var4 }} == {{ var4 }}"

  # commented out: error is raised here (wrongly)
  #- name: print var5, expecting "var5 is <missed>"
  #  debug: msg="var5 is {{ var5 | default('missed') }}"

  - name: include vars from file into var with register (not expanded)
    include_vars:
      file: "include_expansion_vars.yml"
      name: localvar
    register: res

  - name: print localvar.var4 (direct), expecting "a and <c>"
    debug: var=localvar.var4

  - name: print localvar.var4 (host vars), expecting "a and <c>"
    debug: var=hostvars[inventory_hostname].localvar.var4

  - name: print var4 (msg), expecting "a and <c> == a and <c>"
    debug: msg="{{ hostvars[inventory_hostname].localvar.var4 }} == {{ localvar.var4 }}"

  - name: include vars from file into var, without register (expanded only in hostvars)
    include_vars:
      file: "include_expansion_vars.yml"
      name: localres

  - name: print localres.var4 (direct), expecting "a and <c>"
    debug: var=localres.var4

  - name: print localres.var4 (host vars), expecting "a and <c>"
    debug: var=hostvars[inventory_hostname].localres.var4

  - name: print var5, expecting error "var5 is not defined"
    debug: msg="var5 is {{ hostvars[inventory_hostname].localres.var5 | default('missed') }}"

  - name: print localres.var4 (msg), expecting "var4 is a and <c>
    debug: msg="var4 is {{ localres.var4 |default('<missed>') }}"
```
##### EXPECTED RESULTS

`var4` expected to be ok, `var5` must be undefined (and raise error on usage if not defaulted)

##### ACTUAL RESULTS
```
alex@alex-mp (master *=):experiments$ ansible-playbook -l localhost include_expansion_bug.yml

PLAY [localhost] ***********************************************************************************************************************

TASK [include vars from file (mostly ok, broken 'default')] ****************************************************************************
ok: [localhost]

TASK [print var4 (direct), expecting "a and <c>"] **************************************************************************************
ok: [localhost] => {
    "var4": "a and <c>"
}

TASK [print var4 (host vars), expecting "a and <c>"] ***********************************************************************************
ok: [localhost] => {
    "hostvars[inventory_hostname].var4": "a and <c>"
}

TASK [print var4 (msg), expecting "a and <c> == a and <c>"] ****************************************************************************
ok: [localhost] => {
    "msg": "a and <c> == a and <c>"
}

TASK [include vars from file into var with register (not expanded)] ********************************************************************
ok: [localhost]

TASK [print localvar.var4 (direct), expecting "a and <c>"] *****************************************************************************
ok: [localhost] => {
    "localvar.var4": "{{ var1 + ' and ' + var2|default('<c>') }}"
}

TASK [print localvar.var4 (host vars), expecting "a and <c>"] **************************************************************************
ok: [localhost] => {
    "hostvars[inventory_hostname].localvar.var4": "{{ var1 + ' and ' + var2|default('<c>') }}"
}

TASK [print var4 (msg), expecting "a and <c> == a and <c>"] ****************************************************************************
ok: [localhost] => {
    "msg": "{{ var1 + ' and ' + var2|default('<c>') }} == {{ var1 + ' and ' + var2|default('<c>') }}"
}

TASK [include vars from file into var, without register (expanded only in hostvars)] ***************************************************
ok: [localhost]

TASK [print localres.var4 (direct), expecting "a and <c>"] *****************************************************************************
ok: [localhost] => {
    "localres.var4": "VARIABLE IS NOT DEFINED!"
}

TASK [print localres.var4 (host vars), expecting "a and <c>"] **************************************************************************
ok: [localhost] => {
    "hostvars[inventory_hostname].localres.var4": "a and <c>"
}

TASK [print var5, expecting error "var5 is not defined"] *******************************************************************************
ok: [localhost] => {
    "msg": "var5 is {{ var1 + ' and ' + var2 }}"
}

TASK [print localres.var4 (msg), expecting "var4 is a and <c>] *************************************************************************
fatal: [localhost]: FAILED! => {"failed": true, "msg": "the field 'args' has an invalid value, which appears to include a variable that is undefined. The error was: {u'var5': u\"{{ var1 + ' and ' + var2 }}\", u'no_var2': u'b', u'var1': u'a', u'var3': u\"{{ 'just ' + var1 }}\", u'var4': u\"{{ var1 + ' and ' + var2|default('<c>') }}\"}: 'var2' is undefined\n\nThe error appears to have been in '/Users/alex/works/sysadm/ansible-study/experiments/include_expansion_bug.yml': line 53, column 5, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n  - name: print localres.var4 (msg), expecting \"var4 is a and <c>\n    ^ here\n"}

PLAY RECAP *****************************************************************************************************************************
localhost                  : ok=12   changed=0    unreachable=0    failed=1 
```

### LINKED ISSUE #36692: Dynamic vars lookups in templates always contaminate the AnsibleContext
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

lib/ansible/template

##### ANSIBLE VERSION

```
ansible 2.4.2.0
```

##### SUMMARY

Dynamic vars lookups in templates mark the entire Context unsafe, even if no unsafe values were accessed.  This is because AnsibleContext sets itself unsafe whenever a dictionary containing unsafe values/keys is resolved by Jinja2.  The solution is to proxy these dicts instead and set the
context unsafe ONLY if an unsafe value/key is accessed. 

##### STEPS TO REPRODUCE

```yaml
tasks:
  set_fact:
     key_to_safe_dict: "{{ vars['key_to_safe_dict'] }}"
```

##### EXPECTED RESULTS

`vars['keys_to_safe_dict']` is still safe

##### ACTUAL RESULTS

`vars['key_to_safe_dict']` is no longer safe :( 

```

```

### PR DESCRIPTION
**Pull request by DylanYoung**
##### SUMMARY

Fixes #36024: Remove side effects when using `wrap_var`
Fixes #36692: Make the marking of AnsibleContexts as unsafe more granular for dictionaries

These are included in the same PR, as they use similar base functionality (the proxying dict class)

NOTE: to complete this PR, it would be wise to proxy lists in a similar manner to dicts, but this PR is a proof of concept. I'm happy to complete this work if approved.  

NOTE: This PR also removes a duplicate code block.

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

template
set_fact
include_vars
register

##### ANSIBLE VERSION

```
ansible 2.4.2.0
```

##### ADDITIONAL INFORMATION

This is a preliminary (proof of concept) PR.  A few things should change before it's merged:

- AnsibleDictProxy should (maybe) be moved to a different utils module (as the base class itself has no dirct ties to the unsafety system. 

- AnsibleListProxy should be created with analogous functionality

### CODE REVIEW

**Comment by alikins** [comment]
pinging @jimi-c

**Comment by mattclay** [comment]
CI failure in multiple integration tests: https://app.shippable.com/github/ansible/ansible/runs/56486/summary/console

**Comment by DylanYoung** [comment]
On it :)

There are a few gotchas. Just smoothing them out.  

I'm just getting started in ansible, so still not totally familiar with the tooling.

**Comment by gundalow** [comment]
Hi,
The PR should be against the `devel` branch, then if we think it's good we can cherry-pick it into `stable-2.5`

**Comment by DylanYoung** [comment]
Thanks for the tip John! I've been swamped, but when I get a moment, I'll
pull this one down and try to get a proper PR up. All the integration tests
are passing on this branch now, so it shouldn't be too much work from here
:)

**Comment by gundalow** [comment]
That's great, thanks for the update.
This will be for Ansible 2.5.1, so we've got a week or two

**Comment by mattclay** [comment]
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/63441/20/tests

**Comment by DylanYoung** [comment]
Finally got some time to put into this: only 4 months later :p.  Just rerunning tests after rebasing on devel. 

I have a patch done for the corresponding list/sequence functionality, but it still needs to be tested and cleaned up.

**Comment by DylanYoung** [comment]
Looks like there are some CI errors (they're related to pyopen ssl, so I'm going to revert de7983136b9 and try again).  I'm not able to replicate these locally / on my testing machine. Any tips?

File: test.py [hunk 1]
```diff
@@ -0,0 +1,278 @@
+from __future__ import print_function
```
Comment (DylanYoung) [inline test.py]: This file will be removed :)  Though there is an implementation using slots here if that's desired :)

**Comment by DylanYoung** [comment]
Ok.  So I suspect most of the test failures are solved.  I'm a little unsure regarding the JSON RPC (network) failures. 

Does anyone know how this is supposed to work with unsafe values? @jimi-c ? @gundalow ? @mattclay ?

EDIT: Confirmed that the JSON RPC issue is still unresolved.  I suspect there are two options:

1) (IF the RPC is supposed to communicate unsafe values) We should just use the right decoder/encoder pair to properly store the unsafe val. 
2) (IF the RPC is unconcerned with unsafe values) We should implement an encoder that simply resolves the underlying dict (rather than storing an unsafe repr).

I'm not 100% on what these utilities are used for, so just let me know how to proceed (preferably with an explanation). 

Perhaps this comment in lib/ansible/parsing/ajson.py is related:
`# TODO: find way to integrate with the encoding modules do in module_utils`

NOTE: The sanity error is just a circular import. I'm not pushing the fix so I don't overload the CI servers.

**Comment by DylanYoung** [comment]
Lacking response, I've implemented what I consider to be the "safe" solution to the above problem.  It piggybacks on the pickle protocol to serialize, so that adding new unsafe objects will be a simple matter of adding an entry to the AnsibleJSONDecoder.unsafe_cls_map, as long as they implement `reduce_ex` and `__setstate__` (NOTE: this still needs a little work on the Decoder side).

**Comment by DylanYoung** [comment]
Woohoo! We're all passing now.  I have another branch that implements the list version as well. 

There's a bit of cleanup left, but wondering if there's any feedback on the proposed implementation from the core devs before I start cleaning up.

@jimi-c ? @gundalow ? @mattclay ?

**Comment by DylanYoung** [comment]
Any word on this?  I'm happy to finish this up, but would like some indication that this will be merged once it is complete.  Repeatedly rebasing over month long time frames is a pain :p

**Comment by jimi-c** [comment]
Hi @DylanYoung, sorry to take so long to get back to this, however I believe we may have implemented these fixes over the intervening years. If you're interested in verifying that, we'd appreciate otherwise we'll check ourselves to be sure.

**Comment by DylanYoung** [comment]
Of course it still affects me,  lol.

Far as I know this hasn't been implemented,  but a quick check of the
source code should tell you.

**Comment by DylanYoung** [comment]
I'll double check when I get a moment though.   We've already written our playbook to work around this bug.  They're suboptimal and much more verbose than they need to be though and I'd definitely still like to see it fixed.

**Pull request closed without merging**